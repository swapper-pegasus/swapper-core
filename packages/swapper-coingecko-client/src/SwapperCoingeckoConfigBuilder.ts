import qs from 'qs'

import {
  RequestConfigBuilder,
  RequestConfig,
  HttpMethod,
  Params
} from './http/HttpClientInterface'

type Data = Params;

export const THRESHOLD_AVOID_REQUEST_URL_TOO_LARGE = 4096

export default class SwapperCoingeckoConfigBuilder implements RequestConfigBuilder {
  private baseUrl: string
  private headers: object

  constructor ({ baseUrl }: {baseUrl: string;}) {
    this.baseUrl = baseUrl
    this.headers = {}
  }

  public async build (
    method: HttpMethod,
    path: string,
    params: Data,
    options?: { responseType: 'arraybuffer' }
  ) {
    const requestConfig: RequestConfig = {
      method,
      headers: this.headers,
      url: `${this.baseUrl}${path}`,
      ...(options || {})
    }

    switch (method) {
      case 'get': {
        const requestUrl = this.buildRequestUrl(path, params)
        if (requestUrl.length > THRESHOLD_AVOID_REQUEST_URL_TOO_LARGE) {
          return {
            ...requestConfig,
            method: 'post' as const,
            headers: this.headers,
            data: this.buildData(params)
          }
        }
        return {
          ...requestConfig,
          url: requestUrl
        }
      }
      case 'post': {
        return {
          ...requestConfig,
          data: this.buildData(params)
        }
      }
      case 'put': {
        return {
          ...requestConfig,
          data: this.buildData(params)
        }
      }
      case 'delete': {
        const requestUrl = this.buildRequestUrl(
          path,
          this.buildData(params)
        )
        return {
          ...requestConfig,
          url: requestUrl
        }
      }
      default: {
        throw new Error(`${method} method is not supported`)
      }
    }
  }

  private buildRequestUrl (path: string, params: Data): string {
    return `${this.baseUrl}${path}?${qs.stringify(params)}`
  }

  private buildData<T extends Data> (params: T) {
    return params
  }
}
