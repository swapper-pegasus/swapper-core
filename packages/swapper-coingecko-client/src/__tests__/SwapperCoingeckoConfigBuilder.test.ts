import SwapperCoingeckoConfigBuilder, { THRESHOLD_AVOID_REQUEST_URL_TOO_LARGE } from '../SwapperCoingeckoConfigBuilder'

const builder = new SwapperCoingeckoConfigBuilder({ baseUrl: 'https://mi-test-url.com/' })

function generateLongString (length: number): string {
  let result = ''
  for (let i = 0; i < length; i++) {
    result += 'a'
  }
  return result
}

describe('SwapperCoingeckoConfigBuilder', () => {
  describe('get', () => {
    it('success', async () => {
      const result = await builder.build(
        'get',
        'get-path',
        { name: 'name', surname: 'surname' }
      )
      expect(result).toMatchSnapshot()
    })
    it('url too long', async () => {
      const result = await builder.build(
        'get',
        generateLongString(THRESHOLD_AVOID_REQUEST_URL_TOO_LARGE),
        { name: 'name', surname: 'surname' }
      )
      expect(result).toMatchSnapshot()
    })
  })
  it('post', async () => {
    const result = await builder.build(
      'post',
      'post-path',
      { name: 'name', surname: 'surname' }
    )
    expect(result).toMatchSnapshot()
  })
  it('put', async () => {
    const result = await builder.build(
      'put',
      'put-path',
      { name: 'name', surname: 'surname' }
    )
    expect(result).toMatchSnapshot()
  })
  it('delete', async () => {
    const result = await builder.build(
      'delete',
      'delete-path',
      { name: 'name', surname: 'surname' }
    )
    expect(result).toMatchSnapshot()
  })
})
