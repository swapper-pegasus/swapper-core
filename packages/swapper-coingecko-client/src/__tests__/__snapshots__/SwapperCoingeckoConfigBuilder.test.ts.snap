// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapperCoingeckoConfigBuilder delete 1`] = `
Object {
  "headers": Object {},
  "method": "delete",
  "url": "https://mi-test-url.com/delete-path?name=name&surname=surname",
}
`;

exports[`SwapperCoingeckoConfigBuilder get success 1`] = `
Object {
  "headers": Object {},
  "method": "get",
  "url": "https://mi-test-url.com/get-path?name=name&surname=surname",
}
`;

exports[`SwapperCoingeckoConfigBuilder get url too long 1`] = `
Object {
  "data": Object {
    "name": "name",
    "surname": "surname",
  },
  "headers": Object {},
  "method": "post",
  "url": "https://mi-test-url.com/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
}
`;

exports[`SwapperCoingeckoConfigBuilder post 1`] = `
Object {
  "data": Object {
    "name": "name",
    "surname": "surname",
  },
  "headers": Object {},
  "method": "post",
  "url": "https://mi-test-url.com/post-path",
}
`;

exports[`SwapperCoingeckoConfigBuilder put 1`] = `
Object {
  "data": Object {
    "name": "name",
    "surname": "surname",
  },
  "headers": Object {},
  "method": "put",
  "url": "https://mi-test-url.com/put-path",
}
`;
